### PR TITLE
[notebook2, apiserver] mount google bucket, user-jwt secret

### PR DESCRIPTION
**Pull request by akotlar**
This works in-cluster, is currently running. Has a concomitant change in the creation script https://github.com/hail-is/hail/pull/5618/commits/125637488a54d100b992866b6f38437e5814ac28

Still figuring permissions out; the bucket acl command I use to add user permissions is not quite the right approach it appears. I'll fix this shortly in the user-scripts pr, for the meantime I'm updating the permissions manually.
  * You may not have sufficient permissions until I fix this in the user scripts codebase, but that's not an issue that affects the correctness of this pr.

Here it is in action (notice hail-jupyter's /bin is gone):
Screenshot 2019-04-04 17 43 13

Screenshot 2019-04-04 17 48 46

Also includes some modifications @danking wanted from another PR

cc @cseed, @jigold

### CODE REVIEW

File: notebook2/notebook/notebook.py [hunk 1]
```diff
@@ -370,10 +387,9 @@ def notebook_post():
 
     jupyter_token = uuid.uuid4().hex
     name = request.form.get('name', 'a_notebook')
-    safe_user_id = user_id_transform(g.user['auth0_id'])
+    safe_id = user_id_transform(g.user['auth0_id'])
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: name changed to fit in line ending 79 char constraint. prepping for future flake8 linting

**Comment by akotlar** [comment]
Regarding permissions. No problems except when creating folders. It appears to me that the folder is being created as separate bucket, rather than as an object in the bucket:

"""
...
exceptions.from_http_response(response) google.api_core.exceptions.Forbidden: 403 GET https://www.googleapis.com/storage/v1/b/untitled-folder?projection=noAcl: user-nrru16jaxrwmnzkv5f35xfibg@hail-vdc.iam.gserviceaccount.com does not have storage.buckets.get access to **untitled-folder**.
"""

We of course don't want this to happen. It appears that the "solution" is to first click on an already-created folder; this populates the path with /<bucket_name>/<folder>. Back up to ../ and create the folder, voila. 

So this seems like something that could be fixed in jgscm, or potentially by setting "--notebook-dir" in addition to "GoogleStorageContentManager.default_path". ("default_path" is used, because apparently setting ["--notebook-dir" to set the root directory to the chosen bucket doesn't work](https://github.com/src-d/jgscm#usage).

**Comment by akotlar** [comment]
@danking, @cseed An attempt to use --notebook-dir failed (Didn't understand the path). Will make another attempt to set this as a config, but if not, I think we should defer folder creation as an improvement to jgscm, I'll open an issue. Have forked jgscm, and have identified what appears a likely path to the fix.

**Comment by cseed** [comment]
> I think we should defer folder creation as an improvement to jgscm

That's fine with me.  As a workaround, users can create folders with gsutil for the time being.

I think we should just fork it into the hail-is organization, fix it there, install from that repo in our build (don't need to go so far as to publish), and offer changes upstream, which they can take or not.  Sound good?

**Comment by akotlar** [comment]
> > I think we should defer folder creation as an improvement to jgscm
> 
> That's fine with me. As a workaround, users can create folders with gsutil for the time being.
> 
> I think we should just fork it into the hail-is organization, fix it there, install from that repo in our build (don't need to go so far as to publish), and offer changes upstream, which they can take or not. Sound good?

Yep. Shooting to get the fix PRd today.

**Comment by danking** [comment]
Do dataproc Jupyter notebooks also fail to create folders?

**Comment by danking** [comment]
Yeah this definitely works with a cloudtools notebook.

**Comment by danking** [comment]
It seems like we might just have a misconfiguration since whatever cloudtools does works? This is the notebook initialization script: https://github.com/Nealelab/cloudtools/blob/master/cloudtools/init_notebook.py#L164

**Comment by akotlar** [comment]
@dnaking I assume that the sa running cloudyos has read/list/wrote bucket permissions?

If so are you sure that when you create a folder in / it isn’t just creating a new bucket?

**Comment by akotlar** [comment]
In the cloudtools confit it doesn’t actually appear that a bucket is being mounted.

**Comment by cseed** [comment]
The use case is slightly different:
 - the notebooks in cloudtools shows up in the project (it gives the list of buckets), not inside a specfic bucket,
 - in the app, users shouldn't have get buckets, unlike in cloudtools

I think this probably explains the difference.

**Comment by danking** [comment]
The missing permission is `storage.buckets.get` though? It seems reasonable for a user to [be able to read metadata](https://cloud.google.com/storage/docs/access-control/iam-permissions) about their own bucket. I'd wager that jgscm was designed for use with the `roles/storage.legacyBucketWriter` role granted on their bucket. What role are we currently granting?

**Comment by akotlar** [comment]
> The missing permission is `storage.buckets.get` though? It seems reasonable for a user to [be able to read metadata](https://cloud.google.com/storage/docs/access-control/iam-permissions) about their own bucket. I'd wager that jgscm was designed for use with the `roles/storage.legacyBucketWriter` role granted on their bucket. What role are we currently granting?

The problem I believe is that they would need project-wide read/list permissions.

**Comment by akotlar** [comment]
To make this more concrete, all users are legacyBucketOwner on their own bucket (that will be changed to Read/Write). That is not enough.

**Comment by cseed** [comment]
I agree users should have `storage.buckets.get` only to their own folder, and not `storage.buckets.get`.  However, it looks like it is trying to create a bucket with the new folder name, and failing against that (non-existent) bucket:

 > exceptions.from_http_response(response) google.api_core.exceptions.Forbidden: 403 GET https://www.googleapis.com/storage/v1/b/untitled-folder?projection=noAcl: user-nrru16jaxrwmnzkv5f35xfibg@hail-vdc.iam.gserviceaccount.com does not have storage.buckets.get access to untitled-folder.

That's untitled-folder.  Maybe the error is not permissions at all, but it is using the wrong base directory to create the folder?

**Comment by danking** [comment]
What's the full stack trace?

**Comment by akotlar** [comment]
> I agree users should have `storage.buckets.get` only to their own folder, and not `storage.buckets.get`. However, it looks like it is trying to create a bucket with the new folder name, and failing against that (non-existent) bucket:
> 
> > exceptions.from_http_response(response) google.api_core.exceptions.Forbidden: 403 GET https://www.googleapis.com/storage/v1/b/untitled-folder?projection=noAcl: [user-nrru16jaxrwmnzkv5f35xfibg@hail-vdc.iam.gserviceaccount.com](mailto:user-nrru16jaxrwmnzkv5f35xfibg@hail-vdc.iam.gserviceaccount.com) does not have storage.buckets.get access to untitled-folder.
> 
> That's untitled-folder. Maybe the error is not permissions at all, but it is using the wrong base directory to create the folder?

Right, I mentioned this above. It appears to be trying to create, or trying to read, untitled_folder as a bucket. If you trick it into using /your_bucket_name as the cwd, folder creation works.

**Comment by akotlar** [comment]
> What's the full stack trace?

https://github.com/src-d/jgscm/issues/13

**Comment by danking** [comment]
I see, so our proposed fix is this? https://github.com/src-d/jgscm/compare/master...danking:patch-1

**Comment by danking** [comment]
disregard, I think I finally understand the folder shouldn't be a bucket

File: apiserver/Makefile [hunk 2]
```diff
@@ -13,7 +14,11 @@ install-hail-locally:
 	cp -a ../hail/build/libs/hail-all-spark.jar build/hail/jars
 	cp -a ../hail/python/hail build/hail/python
 
-build: install-hail-locally
+jupyter_notebook_config.py:
```
Comment (danking) [inline apiserver/Makefile]: let's also add this to `.gitignore`

File: apiserver/jupyter_notebook_config.py.in [hunk 3]
```diff
@@ -0,0 +1,767 @@
+# Configuration file for jupyter-notebook.
```
Comment (cseed) [inline apiserver/jupyter_notebook_config.py.in]: I'd prefer to strip all the comments, otherwise the active code gets lost in a sea of noise.

**Comment by cseed** [comment]
I noticed a related inconsistency: when I first open the notebook, the path is /.  If I click into a subdirectory, it changes to /user-xx/subdir picking up the bucket.

**Comment by akotlar** [comment]
Yes, I would like to fix that; this was initially what I had hoped —notebook-dir would do.

**Pull request merged**